### ISSUE TITLE
Enhancement (possibly for next version): two more command line options

### ISSUE THREAD

**Issue by reks99** [fesch/Structorizer.Desktop#179]
Hi,

it would be nice to have the following call sequence:
*Structorizer.x [-g language] [nsd-file] [generated-code-file]*

1. First option says for which language code should be generated - if only an nsd file is given, code could be printed to std-out.
2. The nsd-file is, until now, the existing nsd-file, generated by Structorizer.
3. The generated-code-file should be the output file.

Then there wouldn't be the need for a graphical interface to show. This should allow to generate code from an nsd file in batch mode, perhaps with a call from another IDE.

Perhaps the input file can have a *-i* and the output file a *-o* option. 

If only one or no filename, without an option, is given then Structorizer should behave as now.

What do you think about this.
Fine regards
Rolf

**Comment by codemanyak** [comment]
I don't see a fundamental obstacle.
Neither the NSD parser nor the code generators depend on GUI functionality. The existing references (e.g. to request confirmation for overwriting an existing target file) may be bypassed.
One issue are the export options. These would now be taken from the structorizer.ini file. It might be okay as default but it would seem sensible to define additional command line switches to override these export options.

**Comment by codemanyak** [comment]
One possible idea for the export options might be to use upper/lower-case pairs of letters for enabled/disabled toggle options directly to glue to the "-g" in arbitrary order, e.g.

- L/l for the BASIC line numbers (on/off),
- B/b for the block braces (next line / same line),
- C/c for Instructions as comments (on/off),
- T/t for translation / no translation,

such that a code export to Java with block-opening braces forced on same line, instructions not as comments, forced translations, and BASIC line numbers irrelevant might be specified as
```
Structorizer.x -gbcT JavaGenerator ...
```
An alternative would be to use a separate option keyword for each export option with explicit "on" or "off" argument, e.g. (with the current ini keys used)
```
Structorizer.x -g JavaGenerator -genExportBraces on -genExportComments off -genExportnoConversion off ...
```
In addition, we would need an option to specify the encoding, possibly `-encode UTF-8`.
I guess, with `Strucorizer.x` you mean the respective executible (e.g. Structorizer.exe or Structorizer.sh)?

**Comment by codemanyak** [comment]
Instead of "on" and "off" arguments, option markers + and - could be used for the long version, e.g.:
```
Structorizer.x -g JavaGenerator +genExportBraces -genExportComments -genExportnoConversion ...
```

**Comment by reks99** [comment]
@codemanyak 
Hi Kay,

As I can see, you find the idea interesting.

Yes, as you mentioned. Stuctorizer.x stands for the executable Structorizer program.

To clearify, here is my use case:
- While working on a project, I would like to model a program or function structure
- So I would like start Structorizer (perhaps from an "external program" menu entry of the IDE) to sketch the structure
- Then save the nsd and go back to the IDE.
- Later on, when code becomes required, Structorizier should be called in batch mode, to generate the code from the given nsd in the wanted language. Possibly only an option for _no translation_ to overwrite the settings of the init file will be needed to.
- Some IDE can import data from StdOut or will have the code in a file with a name, they generate.

So I have no need for special export options.

That is where the idea came from.

Perhaps different use cases are possible but I have none in my mind now.

Fine regards
Rolf

**Comment by reks99** [comment]
@codemanyak Hi Kay.

I would suggest that there can be more than one input files, so the call may look like
_Structorizer.x [-g cpp] [-t] [-o output-file-name] [input-file1] [input-file2]...[input-file-n]_

The option "-t" can stand for "no translation"

So structorizer can export e.g. all functions in one output file and with a small amount of editing a class file (for c++ in the above example) with its methods is generated.

Fine regards
Rolf

**Comment by codemanyak** [comment]
Okay, I'm preparing a promising prototype.
The command line would have to look like this:

`Structorizer.x [-g generator] [-b] [-c] [-l] [-t] [-e encoding] [-O | -o output-file] [input-file1] [input-file2]...[input-file-n]`
where

- *generator* will have to be either the language name (case-independent) as seen in the export menu or the generator class name (e.g. for Pascal / Delphi it would be either "pascal" or "delphi" or "PasGenerator"). These names are retrieved from the XML configuration file for the generator plugins (the corresponding file extensions are harder to obtain - it would require to produce an instance of every potential generator class in advance). The structorizer command will print a list of accepted generator names ase usage information if called with the -h option (help) or an illegal generator name.
- `-g` must be the first command line argument followed by the generator name (see above), the remaining options (`-b` for block begin in *next* line, `-c` for instructions as comments, `-l` for BASIC line numbers, `-t` for no translations, `-e` for encoding, followed by a charset name, `-O` for standard output, `-o` for output file, followed by a valid file path) may be given in arbitrary order, even mixed with NSD file names.
- *input-file-i* are names/paths of NSD files to be involved in the code generation. Their given order will be preserved in the content of the output file.
- *output-file* is the name of the source code file to be generated. it will be ensured that the created file will get the generator-specific file extension (e. g. `.cpp` for C++). If neither `-o` nor `-O` is given then the output file name would be derived from the file name of the first input file.

On that occasion, I found it sensible to add a batch mode for the reverse direction as well, i.e. for the Pascal input (it's the only working parser for Structorizer by now):

`Structorizer.x [-p parser] [-e encoding] [-o output-file] [input-file]`
where

- *parser* can be "pas" or "pascal"
- *encoding* is a characterset name (for the input file)
- *input-file* is the name or path of a Pascal source file (if more file paths are given all but the first will be ignored).
- *output-file* is the name or path of the NSD file to be derived. If option `-o` isn't given then the output file name will be derived from the input-file name (by replacing its file extension by ".nsd" or by apping ".nsd" if there is no file extension.

A third parameterized form would be (as mentioned above):
`Structorizer.x -h` to obtain a brief command usage information.

The traditional way to call Structorizer just with an optional nsd file name as first command line argument (further args ignored) will of course remain.

**Comment by reks99** [comment]
@codemanyak 
Hi Kay.

The announcement of the prototype looks great!

I have one suggestion: The generator option may be "-e" or "-c" for export, so the option "-g" can be reserved for a later enhancement to export graphics as well.

Further on in the Linux/Unix environment traditionally the option "-" indicates the StdOut path. 

I will enhance the man page (did Bob sent it to you?) according the enhancement.

Fine regards
Rolf

**Comment by codemanyak** [comment]
What about -x for code eXport?
And the proposal of a bare "-" for stdout is fine with me.

**Comment by reks99** [comment]
@codemanyak 
Hi Kay,

It will be fine, and I will modify the man page this way.

Fine regards
Rolf

**Comment by reks99** [comment]
Hi Kay,

A short question: What is the default charset when code is exported (UTF8)?

**Comment by reks99** [comment]
[Structorizer.txt](https://github.com/fesch/Structorizer.Desktop/files/243757/Structorizer.txt)
Hi Kay, hi Bob

I could upload the text file, which is a unix man page for Structorizer - so you need a tool to display man pages.
It's not finished - more key bindings are missing - but I would ask, if the documentation of the parameters for batch mode are interpreted in the proposed way.

Perhaps you can give a feedback about errors with regard to content and grammar.

Fine regards
Rolf

**Comment by codemanyak** [comment]
Hi Rolf,

you asked (I don't now why the markdown-quoted text isn't shown anymore):

> A short question: What is the default charset when code is exported (UTF8)?

I'd say we could favorize UTF-8. Second option would be the system standard charset.

**Comment by reks99** [comment]
Hi Kay,

For me UTF-8 is ok. 
How would you detect the users language settings? In the shells typically a environment variable, called LANG (in uppercase!) holds for example the string "de_DE.UTF-8".

**Comment by reks99** [comment]
[Structorizer.txt](https://github.com/fesch/Structorizer.Desktop/files/244009/Structorizer.txt)

Hi Kay.

I would ask, if the new option "-" for output to StdOut couldn't be canceled, because it is usually clear, that any output goes to StdOut until the option -o filename is set.

This prevent the problem, that a user give both options and expected to get the export in a file and as a terminal output.

I add a new startskript to pass all arguments to structorizer in a linux environment. This may be added in the next version. It needs to be renamed as Structorizer.sh

Fine regards
Rolf

**Comment by reks99** [comment]
Hi Kay,

first of all I will ask, if the option "-" for export to StdOut can be cancled completly, because it is usually clear, that any output goes to StdOut, except the option "-o filename" is given.

Further on we have to change the skript "Structorizer.sh", where "$1" shell becomes "$*" to pass all arguments to structorizer.

Fine regards
Rolf

**Comment by codemanyak** [comment]
@reks99 
First of all: If you want to send a message to both Bob and me, then there is no need to prefix it with our tags, we'll get it anyway.

Next, w.r.t. to the stdout option. Since the command input is not stdin but a file or a set of files and the output is expected to be source code I find it rather unnatural to throw the latter at the standard output. Only in case a shell generator was specified, this would make some sense, I think. (Under Windows not even this.) I would usually expect a file as output. In a way, it is rather a kind of compiler than a filter. And the most convenient thing would be to derive the output file name from the input file name. That's why I thought (and still think) *this* should be the default. Such that for the (IMHO) rather less likely case someone wants to pipe the resulting source code to another program (directly to a compiler?) an option will be necessary to specify standard output as target. By the way, if you want stdout to be the default then the option -o wouldn't make sense, either, because you could simply redirect the output to a file by `>`.

The need to modify the Structorizer.sh was so obvious that I didn't dare to mention it.
@fesch Less obvious is that the Structorizer.exe also needs a modification, of which I'm not sure whether it will happen automatically with a Structorizer.java modification. The strange thing was that under windows I could test the Structorizer.java enhancements only via direct java invocation of the exported jar file while using the existing Structorizer.exe with the same new jar file did not reflect my changes in any way.

**Comment by codemanyak** [comment]
@reks99 
In an earlier comment you asked how I'd find out the user's or system charset setting. Well, I simply would have left this to the Java library. ;-)

**Comment by codemanyak** [comment]
My suggestion for the SYNOPSIS would be:

`Structorizer [ nsd-file ]`
`Structorizer -x generator [ -b ] [ -c ] [ -f ] [ -l ] [ -t ] [ -e encoding ] [ - | -o output-file ] nsd-file ...`
`Structorizer -p parser [ -f ] [ -e encoding ] [ -o output-file ] source-file ...`
`Structorizer -h`
 
Here `-f` might be an option to force overwriting an existing output-file with same name. If not given, Structorizer might create numbered file versions, i.e. `output-file.0.ext`, `output-file.1.ext` etc. (where .ext is the preserved file name extension).
With the `-p` variant, I thought, it could be allowed to specify several source files, their parsing would be done in a loop, deriving the target file names from the respective source file name by replacing the extension with .nsd (while `-o` isn't given). With `-o` given, the file numbering approach in analogy to opion `-f` could also be used (the first file would be `output-file.0.nsd`, then `output-file.1.nsd` etc.)

**Comment by codemanyak** [comment]
Prototype (3.24-11 candidate) is in the delivery pipeline.

**Comment by codemanyak** [comment]
The content of a the `Structorizer.bat` file as recommended to be used for the command-line mode contains two flaws:

1. Some of the file name separators are wrong ('/' may work in Windows, too, but only if there is no backslash in the same path).
2. In order to make it callable from a different drive, the drive extraction should also be added to the jar path.

Hence, the correct content of `Structorizer.bat` would look like this:
```
java -jar %~d0%~p0Structorizer.app\Contents\Java\Structorizer.jar %*
```

**Comment by codemanyak** [comment]
Corrected batch file is due by next version (3.24-12).
